### ISSUE TITLE
bulk insert - some records don't enter the db

### ISSUE THREAD

**Issue by paulbjensen** [TimelordUK/node-sqlserver-v8#4]
Hi,

I don't know the exact cause of this, but we encountered a problem with using the bulk insert feature of this library. The standard query option that we fell back to works like a charm.

We had a CSV file with 13.8m rows in it that we streamed in batches of 25,000 into the tableMgr.insertRows function. At the end of the run we found that the script exited with 13.5m records inserted into the database (SQL Server 2008 R2). We then used a standard query to run through the csv again, and the missing records were inserted into the database.

The strange thing was that records in the very early parts of the csv file had been missed by the bulk insert, which wasn't what I expected.

I can provide supporting files (db schema, csv file, the script I ran) if that would help, with the caveat that the data is representative of the data inserted, but not the actual data.

**Comment by TimelordUK** [comment]
Hi,

sorry to hear about the issue.  Supporting files would be enormously helpful in trying to find the root cause of the problem.   If you can make them available, that would be appreciated.

One interesting thing to try if you had time is to bypass my java script bulk manager and write bulk operations directly.  This is pretty easy to do, rather than hand in a single parameter per entry in an array you would hand in an array of values for each column i.e.   [   c1_val_1, c1_val_2 ... ] , [ c2_val_1, c2_val_2].   i.e. it is an array of arrays where all column values for all rows are represented in the array.  All sub-arrays must there be the same length so a row N would the N'th index from each array. The signature should be the same as if you are writing a single row.  That is all the java script is doing along with batching it up. 

I have to first work out if the error lies within the c++ driver itself or the javascript.   The c++ driver looks to see if a parameter is an array and if so will reserve a vector of parameters rather than a single parameter.  But the code path is very similar in both cases.  The sorts of things I would therefore try is to build say 10 batches of 25k rows and simply send them in one after the other.  If this works, it would point to the javascript that is wrong.

If i can have files that reproduce the issue that would be of tremendous help.

it is good you found a suitable work around at least until the issue has been solved. Thanks for using my library and for taking the time to point out this problem.

Stephen

**Comment by TimelordUK** [comment]
as a follow up, I added a new unit test that creates a data set of 250k rows in 25k chunks 13 columns per row.  I found an issue with the bulk table meta data which could return repeat column descriptions. This caused my test to fail.

I am not sure if this was the cause of your problem, but my test now passes and I can write 250k rows into the database.   

I will close the issue as I consider the functionality to be working at least with the table types I have tested with.  The JS seems to chunk the rows correctly and the c++ is writing the vectors it is given without any loss.

the test is called huge-bulk.cs - it should be possible to use your table description to test if you wish to do so.
